sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,178 @@
+from __future__ import division
+
+__author__ = "Henry Lin <hlin117@gmail.com>"
+
+import numpy as np
+import scipy.sparse as sp
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.utils import column_or_1d, check_array
+from sklearn.utils.validation import check_is_fitted
+
+__all__ = [
+    "Discretizer"
+]
+
+
+class Discretizer(BaseEstimator, TransformerMixin):
+    """Bins continuous data into k equal width intervals.
+
+    Parameters
+    ----------
+    n_bins : int (default=2)
+        The number of bins to produce. The intervals for the bins are
+        determined by the minimum and maximum of the input data.
+        Raises ValueError if n_bins < 2.
+
+    categorical_features : array-like or None (default=None)
+        Specifies the indices of categorical columns which are not to
+        be discretized. If None, assumes that all columns are continuous
+        features.
+
+    Attributes
+    ----------
+    min_ : float
+        The minimum value of the input data.
+
+    max_ : float
+        The maximum value of the input data.
+
+    cut_points_ : array, shape [numBins - 1]
+        Contains the boundaries for which the data lies. Each interval
+        has an open left boundary, and a closed right boundary.
+
+    n_features_ : int
+        The number of features from the original dataset.
+
+    continuous_mask_ : array, shape [n_continuous_features_]
+        An array of booleans, representing columns for which there are
+        continuous features.
+
+    Example
+    -------
+    >>> # X has two examples, with X[:, 2] as categorical
+    >>> X = [[0, 1, 0, 4], \
+             [6, 7, 1, 5]]
+    >>> from sklearn.preprocessing import Discretizer
+    >>> discretizer = Discretizer(n_bins=3, categorical_features=[2])
+    >>> discretizer.fit(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    Discretizer(...)
+    >>> discretizer.cut_points_ # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    array([[ 2. , 3. , 4.3333...],
+           [ 4. , 5. , 4.6666...]])
+    >>> # Transforming X will move the categorical features to the last indices
+    >>> discretizer.transform(X)
+    array([[0, 0, 0, 0],
+           [2, 2, 2, 1]])
+    """
+    sparse_formats = ['csr', 'csc']
+
+    def __init__(self, n_bins=2, categorical_features="all"):
+        if n_bins < 2:
+            raise ValueError("Discretizer received an invalid number "
+                             "of bins")
+        self.n_bins = n_bins
+        self.categorical_features = categorical_features
+
+        # Attributes
+        self.min_ = None
+        self.max_ = None
+        self.cut_points_ = None
+        self.n_features_ = None
+        self.continuous_mask_ = None
+
+    def _check_sparse(self, X, ravel=True):
+        if ravel:
+            return X.toarray().ravel() if sp.issparse(X) else X
+        return X.toarray() if sp.issparse(X) else X
```

Comment:
Neither <a href="https://docs.scipy.org/doc/scipy-0.9.0/reference/generated/scipy.sparse.csr_matrix.html#scipy.sparse.csr_matrix">csr_matrix</a> nor <a href="https://docs.scipy.org/doc/scipy-0.9.0/reference/generated/scipy.sparse.csc_matrix.html#scipy.sparse.csc_matrix">csc_matrix</a> have the min or max function implemented in scipy 0.9.0.

I might take your suggestion in adding a utility in `sparsefuncs_fast.pyx`. Thank you!